### DISCUSSION TITLE
Publishing all maven modules at once

### DISCUSSION THREAD

**Discussion by vkn** [jreleaser/jreleaser#1659] [General]
Hello,
first of all thanks for Jreleaser, it is a great tool!
I was trying to publish a multi-module maven project (a quarkus extension). The parent module was published but the remaining ones failed. Is there an option to publish[ all modules as one archive](https://central.sonatype.org/publish/publish-portal-upload/) so that in case of errors in one of the modules nothing would be published?
Thanks!

**Comment by aalmiray** [comment]
Please provide more details on your setup (Maven? Gradle?). The recently released MavenCentral deployer will create a bundle for all staged artifacts. However, it's up to you to find a way to stage all artifacts as explained in the guide.

**Comment by vkn** [comment]
It is a  maven project containing two modules:

```xml
 <modules>
        <module>deployment</module>
        <module>runtime</module>
    </modules>
```

As far as I understand Jreleaser uploads modules one by one and I'd like to upload them all at once. So if any error happens in a module, nothing would be published.

The plugin config

```xml
  <profile>
            <id>release</id>
            <properties>
                <performRelease>true</performRelease>
            </properties>
            <build>
                <plugins>
                    <plugin>
                        <groupId>org.jreleaser</groupId>
                        <artifactId>jreleaser-maven-plugin</artifactId>
                        <configuration>
                            <jreleaser>
                                <project>
                                    <copyright>The Apache Software License, Version 2.0</copyright>
                                </project>
                                <signing>
                                    <active>ALWAYS</active>
                                    <armored>true</armored>
                                </signing>
                                <deploy>
                                    <maven>
                                        <mavenCentral>
                                            <sonatype>
                                                <active>ALWAYS</active>
                                                <url>https://central.sonatype.com/api/v1/publisher</url>
                                                <stagingRepositories>target/staging-deploy</stagingRepositories>
                                                <retryDelay>120</retryDelay>
                                                <maxRetries>60</maxRetries>
                                            </sonatype>
                                        </mavenCentral>
                                    </maven>
                                </deploy>
                            </jreleaser>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
        </profile>
```

**Reply by aalmiray** [reply]
JReleaser works differently from the deploy and nexus-staging Maven plugins. Those plugins upload artifacts one by one during a run inside the reactor. JReleaser OTOH expects artifacts to be staged to a local directory where it will pick them up and upload them. 

The nexus2 deployer will deploy each staged artifact separately. The mavencentral deployer creates a zip bundle of all staged artifacts. 

You must however stage all artifacts to a local directory, including parent poms, jars, shas. The JReleaser guide shows how to do it https://jreleaser.org/guide/latest/examples/maven/maven-central.html

**Reply by vkn** [reply]
Thanks!. So instead of doing  `deploy` and `jreleaser:deploy` in one step:
`./mvnw -Prelease deploy jreleaser:deploy -DaltDeploymentRepository=local::file:./target/staging-deploy`

it would be better to first stage `./mvnw clean deploy -DaltDeploymentRepository=local::file:./target/staging-deploy` and then release:
`./mvnw -Prelease jreleaser:deploy -DaltDeploymentRepository=local::file:./target/staging-deploy`

**Reply by aalmiray** [reply]
Yes, that is the way. Also, you don't have to specify the alt deployment directory when invoking the jreleaser deploy goal.

**Reply by vkn** [reply]
I've just tried and it worked but jreleaser tried to upload the bundle for each module 3 times in total. Is there a way to avoid it?

**Reply by aalmiray** [reply]
Yes, invoke it once for all modules, usually at the root. Either use `-pl :root-artifactId` or the `-N` flag to stop recursive invocation.
